README.md
# Simplecov::Tdd

A SimpleCov formatter for test driven development. Displays code coverage results in the console for single files

![Example TDD](https://github.com/joshmfrankel/simplecov-tdd/blob/master/example.gif)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'simplecov-tdd'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install simplecov-tdd

## Usage

1. Ensure that you've configured your project to SimpleCov's _Getting Started_ section: https://github.com/colszowka/simplecov#getting-started
2. Set your SimpleCov configuration to the following:

```ruby
require "simplecov/tdd"
SimpleCov.formatter = Simplecov::Formatter::Tdd
# OR use multi-formatter
SimpleCov.formatters = SimpleCov::Formatter::MultiFormatter.new([
  SimpleCov::Formatter::HTMLFormatter,
  Simplecov::Formatter::Tdd
])
```

Simple Setup:

```ruby
# /spec/spec_helper.rb
require "simplecov"
require "simplecov/tdd"
SimpleCov.formatter = SimpleCov::Formatter::Tdd
SimpleCov.start 'rails'
```

3. Run your tests using `rspec path/to/file_spec.rb` or `guard`
4. Fix the missing coverage
5. 💰 Profit! 💰

## Configuration options

There are a few configuration options that may be set before formatting is called.
Generally you may place these after `SimpleCov.formatter` or `SimpleCov.formatters = SimpleCov::Formatter::MultiFormatter.new` in your setup.

### output_style (default: :simple)
When there are lines missing coverage, this option determines how to display output.
The accepted values are `:simple` (default) or `:verbose`.

```ruby
SimpleCov::Formatter::Tdd.output_style = :verbose
```

Here's an example of what :verbose output looks like:

```ruby
app/models/matched_90.rb
90.0% coverage, 167 total lines

The following 2 lines have missing coverage:
[5, 25]

line | source code
-------------------
5 => obj.is_a?(SomeClass)
25 => SomeClass.explode!
```

### debug_mode (default: false)
This is useful for determining if the current file being tested doesn't have
a match from SimpleCov's file list.

```ruby
SimpleCov::Formatter::Tdd.debug_mode = true
```

## Future Features

* Support for minitest

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/joshmfrankel/simplecov-tdd. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## Code of Conduct

Everyone interacting in the Simplecov::Tdd project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/joshmfrankel/simplecov-tdd/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
0.2.0 (2019-03-06)
==================
## Features
* Added style_output configuration option. There are two options: [:simple, :verbose].
:simple is the default, while :verbose outputs line numbers along with source code that
is missing coverage.
* Added debug_mode configuration option for investigating mismatched filenames. Default is disabled

## Improvements
* Improved output of the current tested file to start with the app/ directory instead
of full path
* Slight optimizations done to the way the current filename is detected
* Better documentation added

## Bug fixes
* Files previously containing the phrase 'spec' as part of their name would be
improperly formatted. This caused these files to come up as mismatches.

0.1.0 (2019-02-20)
==================
## Features
* Run code coverage on single spec. Meant to be used with test runners such as guard.
* Output coverage report including current file's lines missing coverage


spec/spec_helper.rb
require "bundler/setup"
require "simplecov"
require "simplecov/tdd"
require "pry"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  config.filter_run focus: true
  config.run_all_when_everything_filtered = true
end


spec/simplecov/tdd_spec.rb
RSpec.describe SimpleCov::Tdd do

  let(:formatter) { SimpleCov::Formatter::Tdd.new }
  let(:matched_90_filename) { "app/models/matched_90.rb" }
  let(:matched_90_filepath) { Dir.pwd + "/" + matched_90_filename }
  let(:matched_90_spec) { "spec/models/matched_90_spec.rb" }
  let(:matched_file_90_coverage) do
    instance_double(
      SimpleCov::SourceFile,
      filename: matched_90_filepath,
      covered_percent: 90.0,
      lines_of_code: 167,
      missed_lines: [
        instance_double(SimpleCov::SourceFile::Line, line_number: 5, src: "    obj.is_a?(SomeClass)"),
        instance_double(SimpleCov::SourceFile::Line, line_number: 25, src: "    SomeClass.explode!")
      ]
    )
  end
  let(:matched_100_filename) { Dir.pwd + "/app/models/matched_100.rb" }
  let(:matched_100_spec) { "spec/models/matched_100_spec.rb" }
  let(:matched_file_100_coverage) do
    instance_double(
      SimpleCov::SourceFile,
      filename: matched_100_filename,
      covered_percent: 100.0,
      lines_of_code: 53,
      missed_lines: []
    )
  end
  let(:unmatched_file) { instance_double(SimpleCov::SourceFile, filename: "spec/models/narp_spec.rb") }
  let(:matched_file_with_intersecting_filename) { "app/models/anotherspec_with_spec.rb" }
  let(:matched_file_with_intersecting_filepath) { Dir.pwd + "/" + matched_file_with_intersecting_filename }
  let(:matched_file_with_intersecting_spec) { "spec/models/anotherspec_with_spec_spec.rb" }
  let(:matched_file_with_intersecting_name_source_file) do
    instance_double(
      SimpleCov::SourceFile,
      filename: matched_file_with_intersecting_filepath,
      covered_percent: 100.0,
      lines_of_code: 53,
      missed_lines: []
    )
  end
  let(:simple_cov_result_stub) do
    instance_double(
      SimpleCov::Result,
      files: [matched_file_90_coverage, unmatched_file, matched_file_100_coverage, matched_file_with_intersecting_name_source_file]
    )
  end

  it "has a version number" do
    expect(SimpleCov::Tdd::VERSION).not_to be nil
  end

  describe "#format" do

    context "with an unmatched filename" do
      it "silently fails" do
        stub_const("ARGV", ["misc", "nope", "2", "spec/unfound_spec.rb"])

        expect {
          formatter.format(simple_cov_result_stub)
        }.not_to raise_error
      end

      context "and debug mode is enabled" do

        it "raises an Exception" do
          stub_const("ARGV", ["misc", "nope", "2", "spec/unfound_spec.rb"])
          SimpleCov::Formatter::Tdd.debug_mode = true

          expect {
            formatter.format(simple_cov_result_stub)
          }.to raise_error(StandardError)
        end
      end
    end

    context "with a matched filename" do

      context "without full 100% coverage" do
        it "displays the coverage overview" do
          stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])

          expect {
            formatter.format(simple_cov_result_stub)
          }.to output(/#{matched_file_90_coverage.covered_percent}% coverage, #{matched_file_90_coverage.lines_of_code} total lines/).to_stdout
        end

        it "displays the missing line number overview" do
          stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])

          expect {
            formatter.format(simple_cov_result_stub)
          }.to output(/The following #{matched_file_90_coverage.missed_lines.size} lines have missing coverage/).to_stdout
        end

        it "displays the specific missing line numbers" do
          stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])
          line_number_array = matched_file_90_coverage.missed_lines.map(&:line_number)

          expect {
            formatter.format(simple_cov_result_stub)
          }.to output(/#{line_number_array}/).to_stdout
        end

        context "and output_style is :verbose" do

          it "displays the line numbers along with matching code" do
            stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])
            SimpleCov::Formatter::Tdd.output_style = :verbose

            expect {
              formatter.format(simple_cov_result_stub)
            }.to output(
              /line \| source code\n-------------------\n#{matched_file_90_coverage.missed_lines.first.line_number} => #{Regexp.quote(matched_file_90_coverage.missed_lines.first.src.strip)}\n#{matched_file_90_coverage.missed_lines[1].line_number} => #{Regexp.quote(matched_file_90_coverage.missed_lines[1].src.strip)}\n/
            ).to_stdout
          end
        end
      end

      context "with full 100% coverage" do
        it "displays the coverage overview" do
          stub_const("ARGV", ["misc", "nope", "2", matched_100_spec])

          expect {
            formatter.format(simple_cov_result_stub)
          }.to output(/#{matched_file_100_coverage.covered_percent}% coverage, #{matched_file_100_coverage.lines_of_code} total lines/).to_stdout
        end

        it "does not display the missing line number overview" do
          stub_const("ARGV", ["misc", "nope", "2", matched_100_spec])

          expect {
            formatter.format(simple_cov_result_stub)
          }.not_to output(/lines have missing coverage/).to_stdout
        end
      end

      context "with different command line argument orders" do
        it "displays the matched filename when first argument" do
          stub_const("ARGV", [matched_90_spec, "misc", "nope", "2"])

          expect {
            formatter.format(simple_cov_result_stub)
          }.to output(/#{matched_90_filename}/).to_stdout
        end

        it "displays the matched filename when middle argument" do
          stub_const("ARGV", ["misc", matched_90_spec, "nope", "2"])

          expect {
            formatter.format(simple_cov_result_stub)
          }.to output(/#{matched_90_filename}/).to_stdout
        end

        it "displays the matched filename when last argument" do
          stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])

          expect {
            formatter.format(simple_cov_result_stub)
          }.to output(/#{matched_90_filename}/).to_stdout
        end
      end

      context "when filename contains the phrase 'spec'" do

        it "displays the matched filename" do
          stub_const("ARGV", [matched_file_with_intersecting_spec, "misc", "nope", "2"])

          expect {
            formatter.format(simple_cov_result_stub)
          }.to output(/#{matched_file_with_intersecting_filename}/).to_stdout
        end
      end
    end
  end
end


spec/simplecov/guard_spec.rb
RSpec.describe Simplecov::Tdd do

  let(:formatter) { Simplecov::Formatter::Tdd.new }
  let(:matched_90_filename) { Dir.pwd + "/app/models/matched_90.rb" }
  let(:matched_90_spec) { "spec/models/matched_90_spec.rb" }
  let(:matched_file_90_coverage) do
    instance_double(
      SimpleCov::SourceFile,
      filename: matched_90_filename,
      covered_percent: 90.0,
      lines_of_code: 167,
      missed_lines: [
        instance_double(SimpleCov::SourceFile::Line, line_number: 5),
        instance_double(SimpleCov::SourceFile::Line, line_number: 25)
      ]
    )
  end
  let(:matched_100_filename) { Dir.pwd + "/app/models/matched_100.rb" }
  let(:matched_100_spec) { "spec/models/matched_100_spec.rb" }
  let(:matched_file_100_coverage) do
    instance_double(
      SimpleCov::SourceFile,
      filename: matched_100_filename,
      covered_percent: 100.0,
      lines_of_code: 53,
      missed_lines: []
    )
  end
  let(:unmatched_file) { instance_double(SimpleCov::SourceFile, filename: "spec/models/narp_spec.rb") }
  let(:simple_cov_result_stub) do
    instance_double(
      SimpleCov::Result,
      files: [matched_file_90_coverage, unmatched_file, matched_file_100_coverage]
    )
  end

  it "has a version number" do
    expect(Simplecov::Tdd::VERSION).not_to be nil
  end

  context "with an unmatched filename" do
    it "silently fails" do
      stub_const("ARGV", ["misc", "nope", "2", "spec/unfound_spec.rb"])

      expect {
        formatter.format(simple_cov_result_stub)
      }.not_to raise_error
    end

    context "and debug mode is enabled" do

      it "raises an Exception" do
        stub_const("ARGV", ["misc", "nope", "2", "spec/unfound_spec.rb"])

        expect {
          formatter.format(simple_cov_result_stub, debug: true)
        }.to raise_error(StandardError)
      end
    end
  end

  context "with a matched filename" do

    context "and without full coverage" do

      it "displays the coverage overview" do
        stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])

        expect {
          formatter.format(simple_cov_result_stub)
        }.to output(/#{matched_file_90_coverage.covered_percent}% coverage, #{matched_file_90_coverage.lines_of_code} total lines/).to_stdout
      end

      it "displays the missing line number overview" do
        stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])

        expect {
          formatter.format(simple_cov_result_stub)
        }.to output(/The following #{matched_file_90_coverage.missed_lines.size} lines have missing coverage/).to_stdout
      end

      it "displays the specific missing line numbers" do
        stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])
        line_number_array = matched_file_90_coverage.missed_lines.map(&:line_number)

        expect {
          formatter.format(simple_cov_result_stub)
        }.to output(/#{line_number_array}/).to_stdout
      end
    end

    context "and with full coverage" do
      it "displays the coverage overview" do
        stub_const("ARGV", ["misc", "nope", "2", matched_100_spec])

        expect {
          formatter.format(simple_cov_result_stub)
        }.to output(/#{matched_file_100_coverage.covered_percent}% coverage, #{matched_file_100_coverage.lines_of_code} total lines/).to_stdout
      end

      it "does not display the missing line number overview" do
        stub_const("ARGV", ["misc", "nope", "2", matched_100_spec])

        expect {
          formatter.format(simple_cov_result_stub)
        }.not_to output(/lines have missing coverage/).to_stdout
      end
    end

    context "with different command line argument orders" do
      it "displays the matched filename when middle argument" do
        stub_const("ARGV", ["misc", matched_90_spec, "nope", "2"])

        expect {
          formatter.format(simple_cov_result_stub)
        }.to output(/#{matched_90_filename}/).to_stdout
      end

      it "displays the matched filename when last argument" do
        stub_const("ARGV", ["misc", "nope", "2", matched_90_spec])

        expect {
          formatter.format(simple_cov_result_stub)
        }.to output(/#{matched_90_filename}/).to_stdout
      end
    end
  end

  # @todo green coverage
  # @todo brown coverage
  # @todo red coverage
  # @todo tty color coverage
end


lib/simplecov/guard.rb
require "simplecov"
require "simplecov/guard/version"

module Simplecov
  module Formatter
    class Guard
      attr_reader :result

      def format(coverage_results)
        @result = single_result(coverage_results)

        puts current_filename
        coverage_overview_text
        missed_line_numbers_message
      end

      private

      def current_filename
        @_current_filename ||= begin
          # @todo ARGV.last is brittle. Is there a regex that could help here?
          Dir.pwd + "/" + ARGV.last.sub("spec", "app").sub("_spec", "")
        end
      end

      def missed_line_numbers
        result.missed_lines.map(&:line_number)
      end

      def missed_line_numbers_message
        return if result.covered_percent == 100

        puts "\nThe following #{missed_line_numbers.size} lines have missing coverage: \n#{missed_line_numbers}"
      end

      def single_result(coverage_results)
        coverage_results.files.select { |file| file.filename == current_filename }[0]
      end

      def coverage_overview_text
        covered_percent = result.covered_percent.round(2)
        overview_text = "#{covered_percent}% coverage, #{result.lines_of_code} total lines"

        colorized_overview_text = if covered_percent > 85
          "\e[32m#{overview_text}\e[0m"
        elsif covered_percent > 65
          "\e[33m#{overview_text}\e[0m"
        else
          "\e[31m#{overview_text}\e[0m"
        end

        puts colorized_overview_text
      end
    end
  end
end


lib/simplecov/tdd.rb
require "simplecov"
require "simplecov/tdd/version"

module SimpleCov
  module Formatter

    # A SimpleCov formatter for generating single file results while using
    # test driven development principles
    class Tdd
      attr_reader :result

      # Hook into SimpleCov format by defining specific implementation
      # of #format for the TDD formatter
      def format(coverage_results)
        @result = single_result(coverage_results)

        # Is there a matching result?
        if result.nil?
          raise StandardError, "Could not find #{current_filename} as part of the SimpleCov::FileList results" if debug_mode
          return # Early return silently dies outside of debug mode
        end

        puts format_filename_for_output
        puts coverage_overview_text
        puts missed_line_numbers_message
      end

      # Class level attr_writer
      # @param style [Symbol] the symbol representation of the output_style
      #   accepted values are: [:simple, :verbose]
      def self.output_style=(style)
        @output_style = style
      end

      # Class level attr_reader
      def self.output_style
        @output_style
      end

      # Instance level method that defers to class_level variable when
      # set via configuration. The default output_style is :simple
      def output_style
        self.class.output_style || :simple
      end

      # Class level attr_writer
      # @param flag [Boolean] the current status (true/false) of debug mode
      def self.debug_mode=(flag)
        @debug_mode = flag
      end

      # Class level attr_reader
      def self.debug_mode
        @debug_mode
      end

      # Instance level method that defers to class_level variable when
      # set via configuration. The default debug_mode is false
      def debug_mode
        self.class.debug_mode || false
      end

      private

      # Match the current file against the SimpleCov::FileList results to determine
      # a matching test
      # @param coverage_results [SimpleCov::Result] the incoming results from running
      #   SimpleCov after tests
      def single_result(coverage_results)
        coverage_results.files.detect { |file| file.filename == current_filename }
      end

      # Infer the current_filename based on incoming ARGV command line
      # arguments and a matching spec type file
      def current_filename
        @_current_filename ||= begin
          # @TODO support for minitest
          argument_index = ARGV.find_index { |arg| arg.match(/_spec.rb/) }

          if argument_index
            Dir.pwd + "/" + ARGV[argument_index].sub("spec/", "app/").sub("_spec.rb", ".rb")
          end
        end
      end

      # Format the currently matched filename to be more human readable
      def format_filename_for_output
        "app/" + current_filename.split("/app/")[1]
      end

      # @todo tty support for colors
      def coverage_overview_text
        covered_percent = result.covered_percent.round(2)
        overview_text = "#{covered_percent}% coverage, #{result.lines_of_code} total lines"
        colorized_overview_text = overview_text

        if STDOUT.tty?
          colorized_overview_text = if covered_percent > 85
            # Green
            "\e[32m#{overview_text}\e[0m"
          elsif covered_percent > 65
            # Orange
            "\e[33m#{overview_text}\e[0m"
          else
            # Red
            "\e[31m#{overview_text}\e[0m"
          end
        end

        colorized_overview_text
      end

      # Display the missing line numbers in a summary message
      def missed_line_numbers_message
        return if result.covered_percent == 100

        "\nThe following #{missed_line_numbers.size} lines have missing coverage: \n\e[31m#{missed_line_numbers}\e[0m".tap do |output|
          output << missed_line_numbers_verbose_message
        end
      end

      # When output_style == :verbose, display a table that includes the
      # line number missing converage as well as the related code
      def missed_line_numbers_verbose_message
        return "" if output_style != :verbose

        "\nline | source code\n-------------------\n".tap do |output|
          result.missed_lines.each do |missed_line|
            output << missed_line.line_number.to_s + " => " + missed_line.src.strip + "\n"
          end
        end
      end

      def missed_line_numbers
        result.missed_lines.map(&:line_number)
      end
    end
  end
end
